### PR TITLE
Es2018 impl

### PR DESCRIPTION
**Pull request by Shilpi3**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -47,11 +47,11 @@
     "normalize-parser-test": "2.0.0",
     "nyc": "13.0.1",
     "regenerate": "^1.4.0",
-    "shift-parser-expectations": "^2017.0.0",
-    "shift-spec": "^2017.0.0",
+    "shift-parser-expectations": "/Users/u326/shift-tools/shift-parser-expectations",
```
Comment (Protryon) [inline package.json]: You probably meant to have a version here.

File: src/parser.js [hunk 2]
```diff
@@ -1687,7 +1722,7 @@ export class GenericParser extends Tokenizer {
       this.expectContextualKeyword('target');
       return this.finishNode(new AST.NewTargetExpression, startState);
     }
-    let callee = this.isolateCoverGrammar(() => this.parseLeftHandSideExpression({ allowCall: false }));
+    let callee = this.isolateCoverGrammar(() => this.parseLeftHandSideExpression({ allowCall: false })); // TO BE DISCUSSED
```
Comment (Protryon) [inline src/parser.js]: Does this still need to be discussed?

File: src/parser.js [hunk 3]
```diff
@@ -2104,25 +2146,66 @@ export class GenericParser extends Tokenizer {
     let startState = this.startNode();
 
     this.lex();
-
-    let properties = [];
+    let rest = null;
+    let properties = [], property = null;
```
Comment (Protryon) [inline src/parser.js]: This `property` declaration probably belongs in the scope of the while loop.

File: src/parser.js [hunk 4]
```diff
@@ -2316,7 +2411,7 @@ export class GenericParser extends Tokenizer {
       this.previousAwaitLocation = previousAwaitLocation;
 
       return {
-        methodOrKey: this.finishNode(new AST.Method({ isAsync: false, isGenerator, name, params, body }), startState),
+        methodOrKey: this.finishNode(new AST.Method({ isAsync, isGenerator, name, params, body }), startState),
```
Comment (Protryon) [inline src/parser.js:2404]: `isAsync` is always false here.

File: src/parser.js [hunk 5]
```diff
@@ -2282,25 +2378,24 @@ export class GenericParser extends Tokenizer {
...
+      this.allowYieldExpression = isGenerator;
+      this.allowAwaitExpression = true;
+      let params = this.parseParams();
+      this.allowYieldExpression = isGenerator;
+      this.allowAwaitExpression = true;
+      let body = this.parseFunctionBody();
+      this.allowYieldExpression = previousYield;
+      this.allowAwaitExpression = previousAwait;
+      return {
+        methodOrKey: this.finishNode(new AST.Method({ isAsync, isGenerator, name, params, body }), startState),
```
Comment (Protryon) [inline src/parser.js:2385]: `isAsync` is always true here.

File: src/parser.js [hunk 6]
```diff
@@ -1857,7 +1894,7 @@ export class GenericParser extends Tokenizer {
     if (this.lookahead.value === 'yield' && this.allowYieldExpression) {
       throw this.createError(ErrorMessages.ILLEGAL_YIELD_IDENTIFIER);
     }
-    if (this.lookahead.value === 'await' && this.allowAwaitExpression) {
+    if (this.lookahead.value === 'await' && (this.allowAwaitExpression)) {
```
Comment (bakkot) [inline src/parser.js]: why this?

File: .gitignore [hunk 7]
```diff
@@ -5,3 +5,8 @@ coverage/
 dist/
 shift-parser.js
 .nyc_output/
+<<<<<<< HEAD
```
Comment (bakkot) [inline .gitignore]: should fix

File: run-my-js.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+var parseScript = require("./dist/index.js").parseScript;
+var parseModule = require("./dist/index.js").parseModule;
+
+var script = `let b = async () => []; for (a in await b());`;
```
Comment (bakkot) [inline run-my-js.js]: I suspect you didn't mean to commit this file?

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "shift-parser",
-  "version": "6.0.1",
+  "version": "7.0.0",
```
Comment (bakkot) [inline package.json]: Leave this out for the moment; we can get the version bump after everything else.

File: src/parser.js [hunk 10]
```diff
@@ -528,7 +528,7 @@ export class GenericParser extends Tokenizer {
         let lexerState = this.saveLexerState();
         this.lex();
         if (!this.hasLineTerminatorBeforeNext && this.match(TokenType.FUNCTION)) {
-          return this.parseFunction({ isExpr: false, inDefault: false, allowGenerator: false, isAsync: true, startState: preAsyncStartState });
+          return this.parseFunction({ isExpr: false, inDefault: true, allowGenerator: true, isAsync: true, startState: preAsyncStartState });
```
Comment (bakkot) [inline src/parser.js]: Why is this `inDefault`?

File: src/parser.js [hunk 11]
```diff
@@ -689,9 +689,13 @@ export class GenericParser extends Tokenizer {
 
   parseForStatement() {
     this.lex();
+    let isAwait = this.eat(TokenType.AWAIT);
```
Comment (bakkot) [inline src/parser.js]: It looks like this would allow `for await` even outside of async functions, which is not legal. Can you ensure that's not the case and add a test to that effect?

File: src/parser.js [hunk 12]
```diff
@@ -1238,6 +1262,11 @@ export class GenericParser extends Tokenizer {
           rest: null,
         }));
       }
+      case 'AssignmentExpression': {
```
Comment (bakkot) [inline src/parser.js]: Why is this here? This should only be allowed in the `transformDestructuringWithDefault` case, I would think.

File: src/parser.js (hunk 6 above)
Comment (Shilpi3) [reply src/parser.js]: I don't know?

File: src/parser.js (hunk 12 above)
Comment (Shilpi3) [reply src/parser.js]: yes. Forgot to remove it.

File: package.json (hunk 9 above)
Comment (bakkot) [reply package.json]: @Shilpi3 this isn't resolved; this PR should leave the version as 6.0.1.

File: src/parser.js (hunk 10 above)
Comment (bakkot) [reply src/parser.js]: @Shilpi3 I still don't know why this is `inDefault`

File: src/parser.js (hunk 12 above)
Comment (bakkot) [reply src/parser.js]: @Shilpi3 it's still not removed

File: src/parser.js (hunk 6 above)
Comment (bakkot) [reply src/parser.js]: @Shilpi3 just remove the extra parentheses so it's how it was.

File: src/parser.js (hunk 3 above)
Comment (bakkot) [reply src/parser.js]: @Shilpi3 I agree with this.

File: test/test262-parser-tests/expectations.js [hunk 13]
```diff
@@ -66,6 +55,14 @@ export default {
       // https://github.com/tc39/test262-parser-tests/issues/15
       '14eaa7e71c682461.js',
       'aca911e336954a5b.js',
+
+      // errors that michael disagrees with (function name `eval`/`arguments`/`package`/`static` invalid for strict mode function bodies)
```
Comment (bakkot) [inline test/test262-parser-tests/expectations.js:59]: Is there an open issue for this? If not you should open one. This comment should refer to an issue, not a person.

File: src/parser.js [hunk 14]
```diff
@@ -2101,27 +2141,52 @@ export class GenericParser extends Tokenizer {
...
+        if (property.type === 'ObjectExpression' || property.type === 'ArrayExpression') {
+          this.isBindingElement = this.isAssignmentTarget = false;
+        }
+        rest = property;
+        properties.push(spreadPropertyOrAssignmentTarget);
+      } else {
+        property = this.inheritCoverGrammar(this.parsePropertyDefinition);
+        properties.push(property);
+      }
+      if (!this.match(TokenType.RBRACE) && !this.match(TokenType.ELLIPSIS)) {
```
Comment (bakkot) [inline src/parser.js]: Are you sure the `&& !this.match(TokenType.ELLIPSIS)` case should be here? It looks like it would allow `({a ...b})` (note absence of comma), which it should not.

File: src/parser.js [hunk 15]
```diff
@@ -2101,27 +2141,52 @@ export class GenericParser extends Tokenizer {
...
-      if (!this.match(TokenType.RBRACE)) {
+      let property = null;
+      let isSpreadProperty = false;
+      let spreadPropertyOrAssignmentTarget = null;
+      if (this.match(TokenType.ELLIPSIS)) {
+        spreadPropertyOrAssignmentTarget = this.parseSpreadPropertyDefinition();
+        isSpreadProperty = true;
+
+        property = spreadPropertyOrAssignmentTarget.expression;
+        if (property.type === 'ObjectExpression' || property.type === 'ArrayExpression') {
```
Comment (bakkot) [inline src/parser.js]: For consistency with `parsePropertyDefinition`, this `if` should be in `parseSpreadPropertyDefinition`.

File: src/parser.js [hunk 16]
```diff
@@ -2245,11 +2311,24 @@ export class GenericParser extends Tokenizer {
...
+    if (isAsync && this.hasLineTerminatorBeforeNext) {
+      isAsync = false;
+      this.restoreLexerState(preAsyncTokenState);
+    }
+
     let isGenerator = !!this.eat(TokenType.MUL);
+    if (isAsync && !this.lookaheadPropertyName()) {
+      isAsync = false;
+      this.restoreLexerState(preAsyncTokenState);
+      isGenerator = !!this.eat(TokenType.MUL);
```
Comment (bakkot) [inline src/parser.js]: If `isAsync` is true, that means the first token after `preAsyncTokenState` is `ASYNC`, i.e. not `MUL`, so you can just put `isGenerator = false;` here (and for consistency move it up a line).

File: src/parser.js [hunk 17]
```diff
@@ -2486,15 +2564,21 @@ export class GenericParser extends Tokenizer {
...
     let properties = [];
+    let rest = null;
     while (!this.match(TokenType.RBRACE)) {
+      if (this.eat(TokenType.ELLIPSIS)) {
+        rest = this.parseBindingIdentifier();
+        break;
+      }
       properties.push(this.parseBindingProperty());
-      if (!this.match(TokenType.RBRACE)) {
+      if (!this.match(TokenType.RBRACE) && !this.match(TokenType.ELLIPSIS)) {
```
Comment (bakkot) [inline src/parser.js]: Same comment on the `&& !this.match(TokenType.ELLIPSIS)` as [above](https://github.com/shapesecurity/shift-parser-js/pull/414/files#r282292413).

File: test/expressions/arrow-expression.js [hunk 18]
```diff
@@ -68,5 +98,15 @@ suite('Parser', () => {
     testParseFailure('() + 0', 'Unexpected token "+"');
     testParseFailure('(10) => 0', 'Illegal arrow function parameter list');
     testParseFailure('(10, 20) => 0', 'Illegal arrow function parameter list');
+    testParseFailure('(...a, b) => {}', ErrorMessages.INVALID_LAST_REST_PARAMETER);
+    testParseFailure('(...a, ...b) => {}', ErrorMessages.INVALID_LAST_REST_PARAMETER);
+    testParseFailure('(a, ...b,) => {}', ErrorMessages.INVALID_LAST_REST_PARAMETER);
+    testParseFailure('(async (...a, b) => {})', ErrorMessages.UNEXPECTED_TOKEN(','));
+    testParseFailure('(async (...a, ...b) => {})', ErrorMessages.UNEXPECTED_TOKEN(','));
```
Comment (bakkot) [inline test/expressions/arrow-expression.js:128]: If they don't already exist, can you add tests confirming that

```js
async (...a, b);
```
and
```js
async (...a, ...b);
```
parse successfully?

File: src/parser.js [hunk 19]
```diff
@@ -1218,10 +1234,20 @@ export class GenericParser extends Tokenizer {
           init: null,
         }));
 
-      case 'ObjectExpression':
+      case 'ObjectExpression': {
+        let last = node.properties.length > 0 ? node.properties[node.properties.length - 1] : void 0;
+        if (last != null && last.type === 'SpreadProperty') {
+          return this.copyNode(node, new AST.ObjectAssignmentTarget({
+            properties: node.properties.slice(0, -1).map(e => e && this.transformDestructuringWithDefault(e)),
+            rest: this.transformDestructuringWithDefault(last.expression),
```
Comment (bakkot) [inline src/parser.js]: Why is this `transformDestructuringWithDefault`? It looks like that would allow `({...a = 0} = 1)`, which it should not. (We should also have a test asserting that that's not allowed.)

File: src/parser.js (hunk 14 above)
Comment (Shilpi3) [reply src/parser.js]: Added a test

File: src/parser.js (hunk 17 above)
Comment (Shilpi3) [reply src/parser.js]: Added a test

File: src/parser.js (hunk 19 above)
Comment (Shilpi3) [reply src/parser.js]: We need this to parse `({a = 0})`, but added a check in `parseObjectExpression` to throw an error if `}` followed by `=`

Comment (bakkot) [reply src/parser.js]: `({a = 0})` should not parse.

File: src/parser.js [hunk 20]
```diff
@@ -2101,27 +2141,52 @@ export class GenericParser extends Tokenizer {
...
     this.lex();
-
+    let rest = null;
     let properties = [];
     while (!this.match(TokenType.RBRACE)) {
-      let property = this.inheritCoverGrammar(this.parsePropertyDefinition);
-      properties.push(property);
+      let property = null;
+      let isSpreadProperty = false;
+      let spreadPropertyOrAssignmentTarget = null;
```
Comment (bakkot) [inline src/parser.js]: This should be inside of the `if` - `let spreadPropertyOrAssignmentTarget = this.parseSpreadPropertyDefinition();`

File: test/miscellaneous/async-await.js [hunk 21]
```diff
@@ -700,22 +768,45 @@ suite('async', () => {
...
+    testParseFailure('async ((a)) => {}', ErrorMessages.UNEXPECTED_TOKEN('=>'));
     testParseFailure('({ async get a(){} })', 'Unexpected identifier');
     testParseFailure('async a => {} ()', 'Unexpected token "("');
     testParseFailure('a + async b => {}', 'Unexpected token "=>"');
     testParseFailure('a + async () => {}', 'Unexpected token "=>"');
     testParseFailure('with({}) async function f(){};', 'Unexpected token "function"');
     testParseFailure('function* a(){ async yield => {}; }', '"yield" may not be used as an identifier in this context');
     testParseFailure('function* a(){ async (yield) => {}; }', 'Unexpected token "=>"');
-    testParseFailure('async function* a(){}', 'Unexpected token "*"');
-    testParseFailure('(async function* (){})', 'Unexpected token "*"');
```
Comment (bakkot) [inline test/miscellaneous/async-await.js:717]: This test and the one above it should be asserted to parse, rather than removed outright.

File: test/statements/expression-statement.js [hunk 22]
```diff
@@ -16,10 +16,37 @@
 
 let testParse = require('../assertions').testParse;
 let stmt = require('../helpers').stmt;
+let testParseFailure = require('../assertions').testParseFailure;
+let ErrorMessages = require('../../src/errors').ErrorMessages;
 
 suite('Parser', () => {
   suite('expression statement', () => {
+    testParse('var {...a} = 0', stmt,
```
Comment (bakkot) [inline test/statements/expression-statement.js:24]: Should also assert that `var {...a.b} = 0` does not parse.

File: test/expressions/object-expression.js [hunk 23]
```diff
@@ -84,6 +86,80 @@ suite('Parser', () => {
...
+          expression: {
+            type: 'ObjectExpression',
+            properties: [],
+          },
+          type: 'SpreadProperty',
+        },
+      ],
+      type: 'ObjectExpression',
+      });
   });
```
Comment (bakkot) [inline test/expressions/object-expression.js:161]: Should also assert that `({...a.b} = 0)` parses successfully.

**Review by bakkot — APPROVED** [review]
